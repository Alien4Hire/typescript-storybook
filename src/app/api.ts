import {
  BaseQueryFn,
  FetchArgs,
  FetchBaseQueryError,
} from '@reduxjs/toolkit/query';

import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { logout, getCaTokenCookie } from './reducers/login/loginSlice';
import { setToastNotification } from './reducers/appSlice/appSlice';

export const baseUrl = process.env.REACT_APP_API || '/api/v1';

declare module '@reduxjs/toolkit/query' {
  interface FetchBaseQueryError {
    status: number;
    data:
    | {
      id?: string;
      location?: string;
      message: string;
      path?: string;
      status?: number;
      type?: string;
    }[]
    | {
      id?: string;
      location?: string;
      message: string;
      path?: string;
      status?: number;
      type?: string;
    };
  }
}

const baseQuery = fetchBaseQuery({
  baseUrl,
  prepareHeaders: (headers, { getState }) => {
    const token = getCaTokenCookie();
    if (token) {
      headers.set('authorization', `Bearer ${token}`);
    }
    headers.set('Content-Type', 'application/json;charset=UTF-8');
    return headers;
  },
});

const baseQueryWithErrorHandling: BaseQueryFn<
  string | FetchArgs,
  unknown,
  FetchBaseQueryError
> = async (args, api, extraOptions) => {
  let result = await baseQuery(args, api, extraOptions);
  if (result.error) {
    if (result.error.status === 401) {
      api.dispatch(logout());
      /**
       * Instead of logging out when the token is invalid, it is possible to refresh the token here.
       */
      // try to get a new token
      // const refreshResult = await baseQuery('/refreshToken', api, extraOptions);
      // if (refreshResult.data) {
      //   // store the new token
      //   api.dispatch(setNewToken(refreshResult.data));
      //   // retry the initial query
      //   result = await baseQuery(args, api, extraOptions);
      // } else {
      //   api.dispatch(logout());
      // }
    } else {
      const message = Array.isArray(result.error?.data)
        ? result.error?.data[0]?.message
        : result.error?.data?.message;
      api.dispatch(
        setToastNotification({
          type: 'error',
          open: true,
          message,
        })
      );
    }
  }
  return result;
};

export const apiBase = createApi({
  reducerPath: 'api',
  baseQuery: baseQueryWithErrorHandling,
  //List of tag associated with our data endpoints
  tagTypes: [
    'Templates',
    'ActivitySettings',
    'TemplateViewTypes',
    'ProductModels',
    'LearningObjectives',
    'Taxonomies',
    'Activities',
    'Lessons',
    'Content',
  ],
  refetchOnFocus: true,
  refetchOnReconnect: true,
  refetchOnMountOrArgChange: true,
  endpoints: () => ({}),
});
