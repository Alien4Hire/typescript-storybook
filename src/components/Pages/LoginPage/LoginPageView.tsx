import React, { useEffect, useRef } from 'react';
import { styled } from '@mui/material/styles';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import Typography from '@mui/material/Typography';
import Snackbar from '@mui/material/Snackbar';
import Alert from '@mui/material/Alert';
import { CaUserT } from '../../../app/reducers/login/loginSlice';

type LoginPageViewT = {
  user: CaUserT | undefined;
  googleClientId: string;
  loadingGoogle: boolean;
  onSuccessHandler: (response: google.accounts.id.CredentialResponse) => void;
  onFailureHandler: (error: any) => void;
};

const StyledLoginWrapper = styled(Card)({
  width: '350px',
  padding: '1em',
  margin: '3em auto',
  'text-align': 'center',
});

const StyledCardActions = styled(CardActions)({
  'justify-content': 'center',
});

const LoginPageView = ({
  user,
  googleClientId,
  loadingGoogle,
  onSuccessHandler,
  onFailureHandler,
}: LoginPageViewT) => {
  const [snackbarOpen, setSnackbarOpen] = React.useState<boolean>(false);

  const handleClose = (
    event?: React.SyntheticEvent | Event,
    reason?: string
  ) => {
    if (reason === 'clickaway') {
      return;
    }
    setSnackbarOpen(false);
  };

  const divRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (divRef?.current && window.google) {
      window.google.accounts.id.initialize({
        client_id: googleClientId,
        callback: (response) => {
          if (response.credential) {
            onSuccessHandler(response);
          } else {
            setSnackbarOpen(true);
            onFailureHandler(new Error('Unable to login'));
          }
        },
      });
      window.google.accounts.id.renderButton(divRef.current, {
        theme: 'outline',
        size: 'medium',
        type: 'standard',
        text: 'signin_with',
      });
    }
  }, [googleClientId, divRef, onSuccessHandler, onFailureHandler]);

  return (
    <>
      <Snackbar
        anchorOrigin={{ horizontal: 'center', vertical: 'bottom' }}
        open={snackbarOpen}
        autoHideDuration={6000}
        onClose={handleClose}
      >
        <Alert onClose={handleClose} severity="error" sx={{ width: '100%' }}>
          Unable to login via Google
        </Alert>
      </Snackbar>
      <StyledLoginWrapper>
        <CardContent>
          <Typography variant="h2" component="div">
            Macmillan Course Assembly
          </Typography>
          <Typography sx={{ mb: 1.5 }} color="text.secondary">
            Please login with your Macmillan Google account.
          </Typography>
        </CardContent>
        <StyledCardActions>
          {loadingGoogle && <p>Please Wait...</p>}
          {!loadingGoogle && <div ref={divRef} />}
        </StyledCardActions>
      </StyledLoginWrapper>
    </>
  );
};

export default LoginPageView;
