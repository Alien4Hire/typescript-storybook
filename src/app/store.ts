import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { combineReducers } from '@reduxjs/toolkit';
import configReducer from './reducers/config/configSlice';
import loginReducer from './reducers/login/loginSlice';
import activitySettingsReducer from './reducers/activitySettings/activitySettingsSlice';
import lessonsReducer from './reducers/lessons/lessonsSlice';
import searchReducer from './reducers/search/searchSlice';
import resourceFilterReducer from './reducers/resourcesFilter/resourcesFilterSlice';
import templateReducer from './reducers/templates/templateSlice';
import appReducer from './reducers/appSlice/appSlice';
import templateViewTypesReducer from './reducers/templateViewTypes/templateViewTypesSlice';
import targetFolderSideBarReducer from './reducers/targetFolderSideBar/targetFolderSideBarSlice';

import { setupListeners } from '@reduxjs/toolkit/query/react';
import { apiBase } from './api';

const rootReducer = combineReducers({
  config: configReducer,
  login: loginReducer,
  activitySettings: activitySettingsReducer,
  lessons: lessonsReducer,
  templates: templateReducer,
  search: searchReducer,
  resourcesFilter: resourceFilterReducer,
  app: appReducer,
  templateViewTypes: templateViewTypesReducer,
  targetFolderSideBar: targetFolderSideBarReducer,
  [apiBase.reducerPath]: apiBase.reducer,
});

// Use also for store setup in jest unit tests
// Use-case: unit test needs preloadedState
// Example: client/src/components/Pages/LoginPage/LoginPage.test.tsx
export const setupStore = (preloadedState: object = {}) => {
  return configureStore({
    reducer: rootReducer,
    devTools: process.env.NODE_ENV !== 'production',
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(apiBase.middleware),
    preloadedState,
  });
};

export const store = setupStore();

setupListeners(store.dispatch);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
