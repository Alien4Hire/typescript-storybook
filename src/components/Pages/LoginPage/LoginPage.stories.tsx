import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import {} from '../../../app/reducers/login/loginSlice';

import LoginPageView from './LoginPageView';

export default {
  title: 'Pages/LoginPage',
  component: LoginPageView,
} as ComponentMeta<typeof LoginPageView>;

const Template: ComponentStory<typeof LoginPageView> = () => (
  <LoginPageView
    user={undefined}
    googleClientId={'mock-google-id'}
    loadingGoogle={false}
    onSuccessHandler={function (
      response: google.accounts.id.CredentialResponse
    ): void {
      throw new Error('Function not implemented.');
    }}
    onFailureHandler={function (error: any): void {
      throw new Error('Function not implemented.');
    }}
  />
);

export const Primary = Template.bind({});
