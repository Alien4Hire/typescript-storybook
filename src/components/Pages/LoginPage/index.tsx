import React from 'react';
import { Navigate, useLocation } from 'react-router-dom';
import CircularProgress from '@mui/material/CircularProgress';
import { styled } from '@mui/material/styles';
import { useAppSelector, useAppDispatch } from '../../../app/hooks';
import LoginPageView from './LoginPageView';
import {
  googleAuthAsync,
  selectUser,
  selectStatus,
} from '../../../app/reducers/login/loginSlice';
import { selectGoogleClientKey } from '../../../app/reducers/config/configSlice';

export const loginPathName = '/login';

const SpinnerWrapper = styled('div')({
  width: '200px',
  margin: '100px auto',
  textAlign: 'center',
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
});

const LoginPage = () => {
  const dispatch = useAppDispatch();
  const user = useAppSelector(selectUser);
  const loginStatus = useAppSelector(selectStatus);
  const { pathname } = useLocation();
  const googleClientId = useAppSelector(selectGoogleClientKey);
  const clientId = googleClientId ?? 'not-configured';

  const onSuccessHandler = (
    response: google.accounts.id.CredentialResponse
  ) => {
    dispatch(
      googleAuthAsync(
        (response as google.accounts.id.CredentialResponse).credential
      )
    );
  };

  const onFailureHandler = (error: any) => {
    console.log('Google login error', error);
  };

  if (user) {
    const navigateTo = pathname === loginPathName ? '/' : pathname;
    return <Navigate to={navigateTo} replace={true} />;
  }

  const verifyingUser = loginStatus === 'loading';
  const loadingGoogle = !googleClientId;

  if (verifyingUser) {
    return (
      <SpinnerWrapper>
        <CircularProgress />
        <span>One moment...</span>
      </SpinnerWrapper>
    );
  }

  return (
    <LoginPageView
      user={user}
      googleClientId={clientId}
      loadingGoogle={loadingGoogle}
      onSuccessHandler={onSuccessHandler}
      onFailureHandler={onFailureHandler}
    />
  );
};

export default LoginPage;
