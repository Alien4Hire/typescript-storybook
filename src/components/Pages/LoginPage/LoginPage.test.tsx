import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';

import LoginPage from './index';
import { setupStore } from '../../../app/store';

const mockStore = setupStore({
  config: {
    status: 'idle',
    googleClientKey: 'mock-client-key',
  },
  login: {
    status: 'idle',
  },
});

const renderComponent = () =>
  render(
    <Provider store={mockStore}>
      <Router>
        <LoginPage />
      </Router>
    </Provider>
  );

describe('<LoginPage />', () => {
  test('renders without crashing', () => {
    const view = renderComponent();
    expect(view).toMatchSnapshot();
  });
});
