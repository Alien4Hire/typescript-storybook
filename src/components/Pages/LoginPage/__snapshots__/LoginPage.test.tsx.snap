// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: #fff;
  color: rgba(0,0,0,0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
}

.c1 {
  overflow: hidden;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
}

.c6 > :not(:first-of-type) {
  margin-left: 8px;
}

.c3 {
  padding: 16px;
}

.c3:last-child {
  padding-bottom: 24px;
}

.c4 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 300;
  font-size: 3.75rem;
  line-height: 1.2;
  -webkit-letter-spacing: -0.00833em;
  -moz-letter-spacing: -0.00833em;
  -ms-letter-spacing: -0.00833em;
  letter-spacing: -0.00833em;
}

.c5 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  color: rgba(0,0,0,0.6);
  margin-bottom: 12px;
}

.c2 {
  width: 350px;
  padding: 1em;
  margin: 3em auto;
  text-align: center;
}

.c7 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<body>
    <div>
      <div
        class="c0 MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 c1 MuiCard-root c2"
      >
        <div
          class="c3 MuiCardContent-root"
        >
          <div
            class="c4 MuiTypography-root MuiTypography-h2"
          >
            Macmillan Course Assembly
          </div>
          <p
            class="c5 MuiTypography-root MuiTypography-body1"
          >
            Please login with your Macmillan Google account.
          </p>
        </div>
        <div
          class="c6 MuiCardActions-root MuiCardActions-spacing c7"
        >
          <div />
        </div>
      </div>
    </div>
  </body>,
  "container": .c0 {
  background-color: #fff;
  color: rgba(0,0,0,0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
}

.c1 {
  overflow: hidden;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
}

.c6 > :not(:first-of-type) {
  margin-left: 8px;
}

.c3 {
  padding: 16px;
}

.c3:last-child {
  padding-bottom: 24px;
}

.c4 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 300;
  font-size: 3.75rem;
  line-height: 1.2;
  -webkit-letter-spacing: -0.00833em;
  -moz-letter-spacing: -0.00833em;
  -ms-letter-spacing: -0.00833em;
  letter-spacing: -0.00833em;
}

.c5 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  color: rgba(0,0,0,0.6);
  margin-bottom: 12px;
}

.c2 {
  width: 350px;
  padding: 1em;
  margin: 3em auto;
  text-align: center;
}

.c7 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
    <div
      class="c0 MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 c1 MuiCard-root c2"
    >
      <div
        class="c3 MuiCardContent-root"
      >
        <div
          class="c4 MuiTypography-root MuiTypography-h2"
        >
          Macmillan Course Assembly
        </div>
        <p
          class="c5 MuiTypography-root MuiTypography-body1"
        >
          Please login with your Macmillan Google account.
        </p>
      </div>
      <div
        class="c6 MuiCardActions-root MuiCardActions-spacing c7"
      >
        <div />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
